node/podcontroller.go

```diff
@@ -313,10 +313,11 @@ func (pc *PodController) Run(ctx context.Context, podSyncWorkers int) (retErr er
 			if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(pod); err != nil {
 				log.G(ctx).Error(err)
 			} else {
+				k8sPod := pod.(*corev1.Pod)
 				pc.knownPods.Delete(key)
 				pc.k8sQ.AddRateLimited(key)
 				// If this pod was in the deletion queue, forget about it
-				pc.deletionQ.Forget(key)
+				pc.deletionQ.Forget(fmt.Sprintf("%v/%v", key, k8sPod.UID))
```

Comment:
Maybe a different approach to stuffing the key in here:

1. You add the pod to the deletionQ
2. The pod gets popped off the deletionQ
3. When the pod gets popped off the deletionQ, you load it from knownPods
4. You check pod.DeletionTimestamp != nil && !running(&pod.Status) -- Maybe even DeletionTimestamp >= time.Now()?

Reply:
Thanks for you suggestions!

But I think pod UID is the real points we concerns. If we did not check uid, we may incorrectly delete a pod.